### DISCUSSION TITLE
Expose API method/interface to execute raw SQL Query

### DISCUSSION THREAD

**Discussion by kevinschaich** [supabase/supabase#11797] [Feature Requests]
Forgive me if I've missed it in the docs, but I think we need to expose a way to write raw SQL via client APIs (to allow access to Postgres functions). Pretty much all database ORMs expose a method to execute a raw SQL Query [(e.g. Prisma)](https://www.prisma.io/docs/concepts/components/prisma-client/raw-database-access#queryraw). While it is currently possible to execute Postgres functions via `let { data, error } = await supabase.rpc(...)`, it becomes impractical at scale and when functions are required within most queries (e.g. when using even out-of-the-box Supabase extensions like Postgis).

For example, I'm using the Postgis extension and the [Finding all data points within a bounding box](https://supabase.com/docs/guides/database/extensions/postgis#finding-all-data-points-within-a-bounding-box) example has been extremely useful. But I have over 100+ geospatial tables that I need to query from my client, which means creating 100+ versions of this function in the currently clunky interface for `table_A`, `table_B`, etc.

A simplified example of the problem:

```sql
select st_asgeojson(geometry) from public.table_A
```

If we had a client-side function which exposes a sql interface, the generalized version which works with all tables could be a one-liner:

```javascript
const select_geojson = async (table) => await supabase.sql(`select st_asgeojson(geometry) from public.${table}`)
```

Thanks for the awesome project 🚀

**Comment by GaryAustin1** [comment]
You have access to the database directly using a standard Postgres connection to run SQL directly.
For example you can connect from node.js with something like https://node-postgres.com/
You need to run SQL from serverside though.

In your example, if that is running on a browser, I as a user can change your SQL to anything I want and make that call.  How would you restrict that?
If you have to lock down the SQL, and move it so the browser can't change it, then it becomes an rpc call. 

I believe there are ways you can pass the table name through to the rpc call and build your query in the function code to use the table name as a variable (probably should screen the table names depending on your function).

**Comment by kevinschaich** [comment]
I'm using it server-side with Node right now but that's a good point @GaryAustin1.

Is there a way for us to add server-only functions to the Javascript API that check the runtime and report an error if executed within client/browser?

**Comment by GaryAustin1** [comment]
Assuming you mean rpc, you can either remove grants on the functions for anon and authenticated (the normal API users) or check for service_role as the current user in SQL in the function.  Then use service_role to make the rpc call on the server side.

**Comment by kiwicopple** [comment]
>If we had a client-side function which exposes a sql interface

Just to follow up on this specific point - there will never be a client-side function which exposes a full-set of SQL to your Postgres instance, because it's a huge security risk. As Gary points out, you have a few options though:

- Use a View or an RPC function: https://supabase.com/docs/reference/javascript/rpc
- Create your own API server and use something like node-postgres:  https://node-postgres.com/
- Use EdgeFuncitons + Deno's `pg` lib: https://github.com/deno-library/pg
- Download your data to the browser and run the SQL there (offline mode, using something like WatermelonDB: https://nozbe.github.io/WatermelonDB/Advanced/Sync.html

I'll convert this one to a discussion for now. There are a few other comments which are similar which you could find useful in there @kevinschaich !

**Comment by kevinschaich** [comment]
Thanks, guys, ended up creating an API route in `next.js` to support this as I ran into a bug (https://github.com/supabase/supabase/issues/11857).

**Comment by kevinschaich** [comment]
@GaryAustin1 @kiwicopple sorry to harp on this, curious if we could make the `.rpc()` function composable as part of the query builder? something like this:

```javascript
let { data, error } = await supabase.from('tableA').select('*').rpc('myFunction')
```

where `myFunction` would be defined as a partial query like this:

```SQL
ST_AsGeoJSON(geometry)
```

and you would expect the full query to be:

```SQL
SELECT *, ST_AsGeoJSON(geometry FROM tableA
```

That would be safe from injection as you're specifying the function code & permissions but way more flexible.

**Reply by steve-chavez** [reply]
@kevinschaich Have you tried using the `geojson()` method? Like:

```js
let { data, error } = await supabase.from('tableA').select('*').geojson()
```

This was added here https://github.com/supabase/postgrest-js/pull/293

**Reply by steve-chavez** [reply]
Another option might be a computed column where inside you use `ST_AsGeoJSON`. There's an example here: https://github.com/supabase/supabase/discussions/3582#discussioncomment-1497587

**Comment by GaryAustin1** [comment]
Not quite sure but this may already be doable with inline functions (they take a setof table type for input and return a table type or setof).
```
    const data = await supabase1
        .from('messages')
        .select('*, echo3')
        .limit(5)
```
With this function:
```
create function echo3(messages)
returns table (
  mytext text
) 
language sql stable
as $$
select $1.message2
$$;
```

Returns 5 rows of the messages table with the message2 column of each row of the messages table as an extra object named echo3.

`....rray":null,"selflink":null,"echo3":"hi"},`

Of course you can do whatever processing you want on the messages columns input and one or more columns out.

I also recommend you consider looking thru and adding to https://github.com/PostgREST/postgrest/issues as any solution like this would get done in that code most likely.

**Comment by iansummerlin** [comment]
alot of people use next.js with supabase it would be nice to do raw sql in api routes 😢
